flush/sink.go

```diff
@@ -0,0 +1,152 @@
+package flush
+
+import (
+	"context"
+	"fmt"
+	"sync/atomic"
+
+	"github.com/uw-labs/substrate"
+	"github.com/uw-labs/substrate-tools/message"
+	"github.com/uw-labs/sync/rungroup"
+)
+
+const (
+	defaultMsgBufferSize = 50
+	defaultAckBufferSize = 100
+)
+
+// An AsyncMessageSinkOption is a function which sets an AsyncMessageSink configuration option.
+type AsyncMessageSinkOption func(ams *AsyncMessageSink)
+
+// WithAckFunc set the AckFunc callback function, which is called for each ack recieved. If the
+// AckFunc returns an error, it will be returned by `Run`.
+func WithAckFunc(ackFn AckFunc) AsyncMessageSinkOption {
+	return func(ams *AsyncMessageSink) {
+		ams.ackFn = ackFn
+	}
+}
+
+// WithMsgBufferSize sets the msg channel buffer size.
+func WithMsgBufferSize(size int) AsyncMessageSinkOption {
+	return func(ams *AsyncMessageSink) {
+		ams.msgBufferSize = size
+	}
+}
+
+// WithAckBufferSize sets the ack channel buffer size.
+func WithAckBufferSize(size int) AsyncMessageSinkOption {
+	return func(ams *AsyncMessageSink) {
+		ams.ackBufferSize = size
+	}
+}
+
+// AckFunc is a callback function executed by `Run` for each message succesfully produced. The ack
+// counter is only incremented once this function returns. If no AckFunc is provided the ack counter
+// is always incremented.
+type AckFunc func(msg substrate.Message) error
+
+// AsyncMessageSink wraps substrate.AsyncMessageSink and provides an interface for interaction with
+// the underlying sink, as well as the capability to flush messages.
+type AsyncMessageSink struct {
+	sink          substrate.AsyncMessageSink
+	msgs          uint64
+	msgBufferSize int
+	msgCh         chan substrate.Message
+	acks          uint64
+	ackBufferSize int
+	ackCh         chan substrate.Message
+	ackFn         AckFunc
+}
+
+// NewAsyncMessageSink returns a pointer a new AsyncMessageSink.
+// See examples/flush/sink.go for example usage.
+func NewAsyncMessageSink(sink substrate.AsyncMessageSink, opts ...AsyncMessageSinkOption) *AsyncMessageSink {
+	ams := &AsyncMessageSink{
+		sink:          sink,
+		msgBufferSize: defaultMsgBufferSize,
+		ackBufferSize: defaultAckBufferSize,
+	}
+
+	for _, opt := range opts {
+		opt(ams)
+	}
+
+	ams.msgCh = make(chan substrate.Message, ams.msgBufferSize)
+	ams.ackCh = make(chan substrate.Message, ams.ackBufferSize)
+
+	return ams
+}
+
+// Run initialises message publishing using the underlying sink and blocks until either an error
+// occurs or the context is done. If an AckFunc is configured, Run will execute it for each ack
+// recieved. If an error is returned, the user should cancel the context to prevent `Flush` from
+// blocking.
+func (ams *AsyncMessageSink) Run(ctx context.Context) error {
+	group, groupctx := rungroup.New(ctx)
+
+	group.Go(func() error {
+		return ams.sink.PublishMessages(groupctx, ams.ackCh, ams.msgCh)
+	})
+
+	group.Go(func() error {
+		for {
+			select {
+			case <-groupctx.Done():
+				return nil
+			case msg := <-ams.ackCh:
```

Comment:
should we be checking for the channel closure before we increment the atomic counter, we could accidentally finish the flush to early as we will mistake the channel closure for an ack. Something like 

``` 
case msg, ok := <- ams.ackCh:
    if !ok {
        return nil
    }
    ...
 ```

Reply:
Don't think this channel get closed any of the current substrate backends, so it's not really necessary...